File: android/app/src/main/java/com/deckboard/app/data/protocol/Messages.kt
package com.deckboard.app.data.protocol

import kotlinx.serialization.Serializable

@Serializable
data class HelloPayload(val role: String, val clientVersion: String)

@Serializable
data class HelloAckPayload(val serverVersion: String, val protocolVersion: Int)

@Serializable
data class PairRequestPayload(val pin: String, val deviceId: String, val deviceName: String)

@Serializable
data class PairResumePayload(val deviceId: String, val deviceToken: String)

@Serializable
data class PairSuccessPayload(val deviceId: String, val deviceToken: String, val deviceName: String)

@Serializable
data class PairErrorPayload(val code: String, val message: String)

@Serializable
data class PairingSessionPayload(val pin: String, val host: String, val port: Int, val qrUri: String, val expiresAt: Long)

@Serializable
data class ButtonPressPayload(val buttonId: String, val label: String, val pressedAt: Long)

@Serializable
data class ButtonPressAckPayload(val buttonId: String, val receivedAt: Long)

@Serializable
data class DeviceConnectedPayload(val deviceId: String, val deviceName: String, val at: Long)

@Serializable
data class DeviceDisconnectedPayload(val deviceId: String, val deviceName: String, val at: Long)

@Serializable
data class ErrorPayload(val code: String, val message: String)

@Serializable
data class LayoutSnapshotPayload(val pages: List<PageConfig>, val activePageId: String)

@Serializable
data class ActionResultPayload(val buttonId: String, val success: Boolean, val message: String? = null)

@Serializable
data class LayoutActivePagePayload(val pageId: String)

@Serializable
data class BackgroundSnapshotPayload(val background: BackgroundConfig)

object MessageType {
    const val HELLO = "hello"
    const val HELLO_ACK = "hello.ack"
    const val PAIR_REQUEST = "pair.request"
    const val PAIR_RESUME = "pair.resume"
    const val PAIR_SUCCESS = "pair.success"
    const val PAIR_ERROR = "pair.error"
    const val PAIRING_SESSION = "pairing.session"
    const val PAIRING_REGENERATE = "pairing.regenerate"
    const val BUTTON_PRESS = "button.press"
    const val BUTTON_PRESS_ACK = "button.press.ack"
    const val DEVICE_CONNECTED = "device.connected"
    const val DEVICE_DISCONNECTED = "device.disconnected"
    const val HEARTBEAT = "heartbeat"
    const val ERROR = "error"
    const val LAYOUT_GET = "layout.get"
    const val LAYOUT_SNAPSHOT = "layout.snapshot"
    const val LAYOUT_UPDATE = "layout.update"
    const val ACTION_RESULT = "action.result"
    const val LAYOUT_ACTIVE_PAGE = "layout.activePage"
    const val BACKGROUND_SNAPSHOT = "background.snapshot"
    const val BACKGROUND_UPDATE = "background.update"
}

sealed class DeckMessage {
    abstract val id: String
    abstract val ts: Long

    data class Hello(override val id: String, override val ts: Long, val payload: HelloPayload) : DeckMessage()
    data class HelloAck(override val id: String, override val ts: Long, val payload: HelloAckPayload) : DeckMessage()
    data class PairRequest(override val id: String, override val ts: Long, val payload: PairRequestPayload) : DeckMessage()
    data class PairResume(override val id: String, override val ts: Long, val payload: PairResumePayload) : DeckMessage()
    data class PairSuccess(override val id: String, override val ts: Long, val payload: PairSuccessPayload) : DeckMessage()
    data class PairError(override val id: String, override val ts: Long, val payload: PairErrorPayload) : DeckMessage()
    data class PairingSession(override val id: String, override val ts: Long, val payload: PairingSessionPayload) : DeckMessage()
    data class PairingRegenerate(override val id: String, override val ts: Long) : DeckMessage()
    data class ButtonPress(override val id: String, override val ts: Long, val payload: ButtonPressPayload) : DeckMessage()
    data class ButtonPressAck(override val id: String, override val ts: Long, val payload: ButtonPressAckPayload) : DeckMessage()
    data class DeviceConnected(override val id: String, override val ts: Long, val payload: DeviceConnectedPayload) : DeckMessage()
    data class DeviceDisconnected(override val id: String, override val ts: Long, val payload: DeviceDisconnectedPayload) : DeckMessage()
    data class Heartbeat(override val id: String, override val ts: Long) : DeckMessage()
    data class Error(override val id: String, override val ts: Long, val payload: ErrorPayload) : DeckMessage()
    data class LayoutSnapshot(override val id: String, override val ts: Long, val payload: LayoutSnapshotPayload) : DeckMessage()
    data class ActionResult(override val id: String, override val ts: Long, val payload: ActionResultPayload) : DeckMessage()
    data class LayoutActivePage(override val id: String, override val ts: Long, val payload: LayoutActivePagePayload) : DeckMessage()
    data class BackgroundSnapshot(override val id: String, override val ts: Long, val payload: BackgroundSnapshotPayload) : DeckMessage()
}

File: shared/src/messages.ts
import { z } from "zod";
import { PROTOCOL_VERSION } from "./constants.js";
import { PageConfigSchema } from "./layout.js";
import { BackgroundConfigSchema } from "./background.js";

// See docs/protocol.md — the hand-maintained wire-format spec. Keep this file
// and that doc in sync; Android's Messages.kt is hand-written to match both.

const HelloPayload = z.object({
  role: z.enum(["controller", "dashboard"]),
  clientVersion: z.string(),
});

const HelloAckPayload = z.object({
  serverVersion: z.string(),
  protocolVersion: z.literal(PROTOCOL_VERSION),
});

const PairRequestPayload = z.object({
  pin: z.string(),
  deviceId: z.string(),
  deviceName: z.string(),
});

const PairResumePayload = z.object({
  deviceId: z.string(),
  deviceToken: z.string(),
});

const PairSuccessPayload = z.object({
  deviceId: z.string(),
  deviceToken: z.string(),
  deviceName: z.string(),
});

const PairErrorCode = z.enum([
  "invalid_pin",
  "expired",
  "already_used",
  "invalid_token",
]);

const PairErrorPayload = z.object({
  code: PairErrorCode,
  message: z.string(),
});

const PairingSessionPayload = z.object({
  pin: z.string(),
  host: z.string(),
  port: z.number(),
  qrUri: z.string(),
  expiresAt: z.number(),
});

const PairingRegeneratePayload = z.object({});

const ButtonPressPayload = z.object({
  buttonId: z.string(),
  label: z.string(),
  pressedAt: z.number(),
});

const ButtonPressAckPayload = z.object({
  buttonId: z.string(),
  receivedAt: z.number(),
});

const DeviceConnectedPayload = z.object({
  deviceId: z.string(),
  deviceName: z.string(),
  at: z.number(),
});

const DeviceDisconnectedPayload = z.object({
  deviceId: z.string(),
  deviceName: z.string(),
  at: z.number(),
});

const HeartbeatPayload = z.object({});

const ErrorPayload = z.object({
  code: z.string(),
  message: z.string(),
});

const LayoutGetPayload = z.object({});

const LayoutSnapshotPayload = z.object({
  pages: z.array(PageConfigSchema),
  activePageId: z.string(),
});

const LayoutUpdatePayload = z.object({
  pages: z.array(PageConfigSchema),
});

const ActionResultPayload = z.object({
  buttonId: z.string(),
  success: z.boolean(),
  message: z.string().optional(),
});

const LayoutActivePagePayload = z.object({
  pageId: z.string(),
});

const BackgroundSnapshotPayload = z.object({
  background: BackgroundConfigSchema,
});

const BackgroundUpdatePayload = z.object({
  background: BackgroundConfigSchema,
});

function envelope<Type extends string, Payload extends z.ZodTypeAny>(
  type: Type,
  payload: Payload,
) {
  return z.object({
    v: z.literal(PROTOCOL_VERSION),
    type: z.literal(type),
    id: z.string(),
    ts: z.number(),
    payload,
  });
}

export const DeckMessageSchema = z.discriminatedUnion("type", [
  envelope("hello", HelloPayload),
  envelope("hello.ack", HelloAckPayload),
  envelope("pair.request", PairRequestPayload),
  envelope("pair.resume", PairResumePayload),
  envelope("pair.success", PairSuccessPayload),
  envelope("pair.error", PairErrorPayload),
  envelope("pairing.session", PairingSessionPayload),
  envelope("pairing.regenerate", PairingRegeneratePayload),
  envelope("button.press", ButtonPressPayload),
  envelope("button.press.ack", ButtonPressAckPayload),
  envelope("device.connected", DeviceConnectedPayload),
  envelope("device.disconnected", DeviceDisconnectedPayload),
  envelope("heartbeat", HeartbeatPayload),
  envelope("error", ErrorPayload),
  envelope("layout.get", LayoutGetPayload),
  envelope("layout.snapshot", LayoutSnapshotPayload),
  envelope("layout.update", LayoutUpdatePayload),
  envelope("action.result", ActionResultPayload),
  envelope("layout.activePage", LayoutActivePagePayload),
  envelope("background.snapshot", BackgroundSnapshotPayload),
  envelope("background.update", BackgroundUpdatePayload),
]);

export type DeckMessage = z.infer<typeof DeckMessageSchema>;

export type DeckMessageType = DeckMessage["type"];

export type DeckMessageOf<Type extends DeckMessageType> = Extract<
  DeckMessage,
  { type: Type }
>;

export type HelloPayload = z.infer<typeof HelloPayload>;
export type HelloAckPayload = z.infer<typeof HelloAckPayload>;
export type PairRequestPayload = z.infer<typeof PairRequestPayload>;
export type PairResumePayload = z.infer<typeof PairResumePayload>;
export type PairSuccessPayload = z.infer<typeof PairSuccessPayload>;
export type PairErrorPayload = z.infer<typeof PairErrorPayload>;
export type PairingSessionPayload = z.infer<typeof PairingSessionPayload>;
export type PairingRegeneratePayload = z.infer<
  typeof PairingRegeneratePayload
>;
export type ButtonPressPayload = z.infer<typeof ButtonPressPayload>;
export type ButtonPressAckPayload = z.infer<typeof ButtonPressAckPayload>;
export type DeviceConnectedPayload = z.infer<typeof DeviceConnectedPayload>;
export type DeviceDisconnectedPayload = z.infer<
  typeof DeviceDisconnectedPayload
>;
export type HeartbeatPayload = z.infer<typeof HeartbeatPayload>;
export type ErrorPayload = z.infer<typeof ErrorPayload>;
export type LayoutGetPayload = z.infer<typeof LayoutGetPayload>;
export type LayoutSnapshotPayload = z.infer<typeof LayoutSnapshotPayload>;
export type LayoutUpdatePayload = z.infer<typeof LayoutUpdatePayload>;
export type ActionResultPayload = z.infer<typeof ActionResultPayload>;
export type LayoutActivePagePayload = z.infer<typeof LayoutActivePagePayload>;
export type BackgroundSnapshotPayload = z.infer<typeof BackgroundSnapshotPayload>;
export type BackgroundUpdatePayload = z.infer<typeof BackgroundUpdatePayload>;

/** Builds a fully-formed envelope for `type`, generating `id`/`ts`/`v`. */
export function makeMessage<Type extends DeckMessageType>(
  type: Type,
  payload: DeckMessageOf<Type>["payload"],
): DeckMessageOf<Type> {
  return {
    v: PROTOCOL_VERSION,
    type,
    id: crypto.randomUUID(),
    ts: Date.now(),
    payload,
  } as DeckMessageOf<Type>;
}
